File: python_pid/zoh_pid.py
"""Zero-order hold (ZOH) discretization for measurement filter.

This module provides the ZOH discretization function for computing
discrete-time filter parameters used in the PID controller's
measurement filter.
"""

import math


def zoh_Fy(TfTs, Tx=1.0, exp=math.exp):
    """Compute filter parameters using zero-order hold discretization.

    Args:
        TfTs: Filter time constant as a multiple of nominal sample
              time
        Tx: Execution period (normalized, default: 1.0)
        exp: Exponential function (default: math.exp)

    Returns:
        tuple: Six state-space matrix coefficients
               (a11, a12, a21, a22, b1, b2)
    """
    # Help variables
    h1 = Tx / TfTs
    h2 = exp(-h1)
    h3 = h1 * h2
    h4 = h3 / TfTs

    # Filter parameters
    a11 = h2 + h3
    a12 = h2
    a21 = -h4
    a22 = h2 - h3
    b1 = 1 - h2 - h3
    b2 = h4

    return a11, a12, a21, a22, b1, b2

File: cpp_pid/zoh_pid.cpp
/**
 * @file zoh_pid.cpp
 * @brief Implementation of ZOH discretization for measurement filter
 */

#include "zoh_pid.h"

FilterParams zoh_Fy(double TfTs, double Tx) {
    // Help variables
    double h1 = Tx / TfTs;
    double h2 = exp(-h1);
    double h3 = h1 * h2;
    double h4 = h3 / TfTs;

    // Filter parameters
    FilterParams params;
    params.a11 = h2 + h3;
    params.a12 = h2;
    params.a21 = -h4;
    params.a22 = h2 - h3;
    params.b1 = 1.0 - h2 - h3;
    params.b2 = h4;

    return params;
}
